### ISSUE TITLE
azure-cli-core: Test failures with conflicting subparsers in Python 3.11.0b3

### ISSUE THREAD

**Issue by major** [Azure/azure-cli#23015]
**Related command**
Attempting to run tests for `azure-cli-core`.

**Describe the bug**
While running `pytest` to test `azure-cli-core`, many tests fail in Python 3.11.0b3 due to conflicting subparsers. Here is an example:

```
__________________ TestParser.test_parser_error_spellchecker ___________________
[gw0] linux -- Python 3.11.0 /usr/bin/python3

self = <azure.cli.core.tests.test_parser.TestParser testMethod=test_parser_error_spellchecker>

    @mock.patch('importlib.import_module', _mock_import_lib)
    @mock.patch('pkgutil.iter_modules', _mock_iter_modules)
    @mock.patch('azure.cli.core.commands._load_command_loader', _mock_load_command_loader)
    @mock.patch('azure.cli.core.extension.get_extension_modname', _mock_extension_modname)
    @mock.patch('azure.cli.core.extension.get_extensions', _mock_get_extensions)
    def test_parser_error_spellchecker(self):
        cli = DummyCli()
        main_loader = MainCommandsLoader(cli)
        cli.loader = main_loader
    
        cli.loader.load_command_table(None)
    
        parser = cli.parser_cls(cli)
>       parser.load_command_table(cli.loader)

src/azure-cli-core/azure/cli/core/tests/test_parser.py:206: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/azure-cli-core/azure/cli/core/parser.py:100: in load_command_table
    command_parser = subparser.add_parser(command_verb,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = _SubParsersAction(option_strings=[], dest='_subcommand', nargs='A...', const=None, default=None, type=None, choices={'module': 'module'}, required=True, help=None, metavar=None)
name = 'module'
kwargs = {'_command_source': 'azure.cli.core.tests.test_parser', 'cli_help': None, 'conflict_handler': 'error', 'description': None, ...}
aliases = ()

    def add_parser(self, name, **kwargs):
        # set prog from the existing prefix
        if kwargs.get('prog') is None:
            kwargs['prog'] = '%s %s' % (self._prog_prefix, name)
    
        aliases = kwargs.pop('aliases', ())
    
        if name in self._name_parser_map:
>           raise ArgumentError(self, _('conflicting subparser: %s') % name)
E           argparse.ArgumentError: argument _subcommand: conflicting subparser: module

/usr/lib64/python3.11/argparse.py:1185: ArgumentError
```

For the full output, check the [build log from a Fedora package build](https://kojipkgs.fedoraproject.org/work/tasks/8140/88698140/build.log). I've also copied out just the pytest output into a [GitHub Gist](https://gist.github.com/major/f7d0de0786f3795056cb030edb4a69cf) to make it easier to read.

**To Reproduce**
Run `pytest src/azure-cli-core` with Python 3.11.0b3.

**Expected behavior**
I expected the tests to pass (as they do in Python 3.10.5).

**Environment summary**
Building an RPM package for `azure-cli` with Python 3.11.0b3 in Fedora 37.

**Comment by yonzhan** [comment]
@jiasli for awareness

**Comment by jiasli** [comment]
Azure CLI currently doesn't support Python 3.11.0b3 yet. We will revisit this topic once Python 3.11 is GA.

**Comment by hroncok** [comment]
argparse.ArgumentError: argument _subcommand: conflicting subparser: module

This usually means there is a problem in the code that was silently ignored until 3.11.

**Comment by mdboom** [comment]
(Greetings -- on the team working on Python 3.11 :wave:)

This seems to do the trick for me.  The workaround for that bug no longer seems to be required.

```diff
diff --git a/src/azure-cli-core/azure/cli/core/parser.py b/src/azure-cli-core/azure/cli/core/parser.py
index fbf0a1ab0..dce310304 100644
--- a/src/azure-cli-core/azure/cli/core/parser.py
+++ b/src/azure-cli-core/azure/cli/core/parser.py
@@ -4,6 +4,7 @@
 # --------------------------------------------------------------------------------------------

 import difflib
+import sys

 import argparse
 import argcomplete
@@ -92,7 +93,8 @@ class AzCliCommandParser(CLICommandParser):
             command_verb = command_name.split()[-1]
             # To work around http://bugs.python.org/issue9253, we artificially add any new
             # parsers we add to the "choices" section of the subparser.
-            subparser.choices[command_verb] = command_verb
+            if sys.version_info < (3, 11):
+                subparser.choices[command_verb] = command_verb

             # inject command_module designer's help formatter -- default is HelpFormatter
             fc = metadata.formatter_class or argparse.HelpFormatter
```

**Comment by major** [comment]
Thanks, @mdboom! I'm giving this a try on 3.11.0rc2 in [Fedora's COPR](https://copr.fedorainfracloud.org/coprs/mhayden/azure-cli/build/4889463/). A quick test on my system looks good.

**Comment by jacoborus** [comment]
> Azure CLI currently doesn't support Python 3.11.0b3 yet. We will revisit this topic once Python 3.11 is GA.

Python 3.11 is out, can you please merge https://github.com/Azure/azure-cli/pull/24109?

**Comment by JoranDox** [comment]
I've got version 2.43 (fresh install of python 3.11 a couple weeks ago, fresh install of az cli today, pip says 2.43, can't run az version because that crashes already) and @mdboom's change seems to be reverted?

```
     92             command_verb = command_name.split()[-1]
     93             # To work around http://bugs.python.org/issue9253, we artificially add any new
     94             # parsers we add to the "choices" section of the subparser.
     95             subparser.choices[command_verb] = command_verb
```

anyway applying the same change manually works, so not sure what to do, but this issue should be reopened I think.

(now az version gives 
```
{
  "azure-cli": "2.43.0",
  "azure-cli-core": "2.43.0",
  "azure-cli-telemetry": "1.0.8",
  "extensions": {}
}
```
)

**Comment by peruzzof** [comment]
Actually the fixed version wasnt released yet, it is planned for the 2.44.0 (planned to be released next week, I guess)

**Comment by JoranDox** [comment]
Ah my bad then, I had understood this to be fixed in 2.40 per @major's comment up above: https://github.com/Azure/azure-cli/issues/23015#issuecomment-1268618199

**Comment by jiasli** [comment]
@peruzzof is right. This change will be included in 2.44.0. You may check the release schedule at https://github.com/Azure/azure-cli/milestone/125
